pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/decoder/protobufnative/PulsarProtobufNativeColumnDecoder.java

```diff
@@ -0,0 +1,386 @@
+
+/**
+ * Pulsar {@link org.apache.pulsar.common.schema.SchemaType#PROTOBUF_NATIVE} ColumnDecoder.
+ */
+public class PulsarProtobufNativeColumnDecoder {
+    private static final Set<Type> SUPPORTED_PRIMITIVE_TYPES = ImmutableSet.of(
+            BooleanType.BOOLEAN,
+            IntegerType.INTEGER,
+            BigintType.BIGINT,
+            RealType.REAL,
+            DoubleType.DOUBLE,
+            VarbinaryType.VARBINARY);
+
+    private final Type columnType;
+    private final String columnMapping;
+    private final String columnName;
+
+    public PulsarProtobufNativeColumnDecoder(DecoderColumnHandle columnHandle) {
+        try {
+            requireNonNull(columnHandle, "columnHandle is null");
+            this.columnType = columnHandle.getType();
+            this.columnMapping = columnHandle.getMapping();
+            this.columnName = columnHandle.getName();
+            checkArgument(!columnHandle.isInternal(),
+                    "unexpected internal column '%s'", columnName);
+            checkArgument(columnHandle.getFormatHint() == null,
+                    "unexpected format hint '%s' defined for column '%s'", columnHandle.getFormatHint(), columnName);
+            checkArgument(columnHandle.getDataFormat() == null,
+                    "unexpected data format '%s' defined for column '%s'", columnHandle.getDataFormat(), columnName);
+            checkArgument(columnHandle.getMapping() != null,
+                    "mapping not defined for column '%s'", columnName);
+            checkArgument(isSupportedType(columnType),
+                    "Unsupported column type '%s' for column '%s'", columnType, columnName);
+        } catch (IllegalArgumentException e) {
+            throw new PrestoException(GENERIC_USER_ERROR, e);
+        }
+    }
+
+    private boolean isSupportedType(Type type) {
+        if (isSupportedPrimitive(type)) {
+            return true;
+        }
+
+        if (type instanceof ArrayType) {
+            checkArgument(type.getTypeParameters().size() == 1,
+                    "expecting exactly one type parameter for array");
+            return isSupportedType(type.getTypeParameters().get(0));
+        }
+
+        if (type instanceof MapType) {
+            List<Type> typeParameters = type.getTypeParameters();
+            checkArgument(typeParameters.size() == 2,
+                    "expecting exactly two type parameters for map");
+            return isSupportedType(type.getTypeParameters().get(1)) && isSupportedType(type.getTypeParameters().get(0));
+        }
+
+        if (type instanceof RowType) {
+            for (Type fieldType : type.getTypeParameters()) {
+                if (!isSupportedType(fieldType)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private boolean isSupportedPrimitive(Type type) {
+        return type instanceof VarcharType || SUPPORTED_PRIMITIVE_TYPES.contains(type);
```

Comment:
Could we add the `VarcharType` to the set `SUPPORTED_PRIMITIVE_TYPES`?

Reply:
I think the current implementation has two advantages :

- VarcharType is the most **frequently** type.
So in most cases, `SUPPORTED_PRIMITIVE_TYPES.contains()` will not be called, current implementation is more efficient.
- VarcharType is the **necessary base** type .
All the type( row,array... ) can print as string in sqlClient, VarcharType is necessary instead of optional.